File: GraffiTab-iOS/Classes/ViewControllers/Main/Main/Home/HomeViewController.m
//
//  HomeViewController.m
//  GraffiTab-iOS
//
//  Created by Georgi Christov on 26/11/2014.
//  Copyright (c) 2014 GraffiTab. All rights reserved.
//

#import "HomeViewController.h"
#import "UIBarButtonItem+Badge.h"
#import "NewestViewController.h"
#import "PopularViewController.h"
#import "HomeStreamViewController.h"

@interface HomeViewController ()

@end

@implementation HomeViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    [self setupTopBar];
    [self setupViewPager];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    if (self.navigationController.isNavigationBarHidden)
        [self.navigationController setNavigationBarHidden:NO animated:NO];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    if (![[Settings getInstance] showedTipMenu])
        [[SlideNavigationController sharedInstance] bounceMenu:MenuLeft withCompletion:nil];
    
    [[Settings getInstance] setShowedTipMenu];
    
    [self loadUnseenNotificationCount];
}

- (void)dealloc {
    NSLog(@"DEALLOC %@", self.class);
}

- (void)onClickToggleMenu {
    [[SlideNavigationController sharedInstance] toggleLeftMenu];
}

- (void)onClickSpray {
    [self performSegueWithIdentifier:@"SEGUE_DRAW" sender:nil];
}

#pragma mark - Loading notifications

- (void)loadUnseenNotificationCount {
    [GTNotificationManager getUnseenNotificationsWithSuccessBlock:^(GTResponseObject *response) {
        self.unseenNotificationsCount = [response.object intValue];
        
        [GTConversationManager getUnseenConversationsCountWithSuccessBlock:^(GTResponseObject *response) {
            self.unseenMessagesCount = [response.object intValue];
            
            [self updateUnseenNotificationsBadge];
        } failureBlock:^(GTResponseObject *response) {}];
    } failureBlock:^(GTResponseObject *response) {}];
}

- (void)updateUnseenNotificationsBadge {
    int totalNotifications = self.unseenMessagesCount + self.unseenNotificationsCount;
    
    self.navigationItem.leftBarButtonItem.badgeValue = totalNotifications > 0 ? [@(totalNotifications) stringValue] : nil;
    self.navigationItem.leftBarButtonItem.badgeBGColor = UIColorFromRGB(COLOR_ORANGE);
    
    [[NSNotificationCenter defaultCenter] postNotificationName:NOTIFICATION_UPDATE_UNSEEN_ITEMS object:nil];
}

#pragma mark - SlideNavigationControllerDelegate

- (BOOL)slideNavigationControllerShouldDisplayLeftMenu {
    return YES;
}

#pragma mark - Setup

- (void)setupTopBar {
    UIImage *image = [[UIImage imageNamed:@"menu.png"] imageWithTint:self.navigationController.navigationBar.tintColor];
    UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
    button.frame = CGRectMake(0, 0, image.size.width, image.size.height);
    [button setImage:image forState:UIControlStateNormal];
    [button addTarget:self action:@selector(onClickToggleMenu) forControlEvents:UIControlEventTouchUpInside];
    
    UIBarButtonItem *navLeftButton = [[UIBarButtonItem alloc] initWithCustomView:button];
    [SlideNavigationController sharedInstance].leftBarButtonItem = navLeftButton;
    
    UIBarButtonItem *create = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"spray.png"] style:UIBarButtonItemStylePlain target:self action:@selector(onClickSpray)];
    self.navigationItem.rightBarButtonItems = @[create];
}

- (void)setupViewPager {
    NSArray *titles = [NSArray arrayWithObjects:@"Home", @"Popular", @"Newest", nil];
    NSMutableArray *viewControllers = [NSMutableArray new];
    
    for (int i = 0; i < titles.count; i++) {
        UIViewController *vc = nil;
        UIStoryboard *mainStoryboard = [SlideNavigationController sharedInstance].storyboard;
        
        if (i == 0) {
            HomeStreamViewController *v = [mainStoryboard instantiateViewControllerWithIdentifier:@"HomeStreamViewController"];
            vc = v;
        }
        else if (i == 1) {
            PopularViewController *v = [mainStoryboard instantiateViewControllerWithIdentifier:@"PopularViewController"];
            vc = v;
        }
        else {
            NewestViewController *v = [mainStoryboard instantiateViewControllerWithIdentifier:@"NewestViewController"];
            vc = v;
        }
        
        [viewControllers addObject:vc];
    }
    
    self.viewControllers = viewControllers;
    self.didChangedPageCompleted = ^(NSInteger cuurentPage, NSString *title) {};
    
    [self reloadData];
}

@end

File: GraffiTab/Classes/ViewControllers/Main/Main/HomeViewController.swift
//
//  HomeViewController.swift
//  GraffiTab
//
//  Created by Georgi Christov on 06/04/2016.
//  Copyright © 2016 GraffiTab. All rights reserved.
//

import UIKit
import CarbonKit
import BBBadgeBarButtonItem
import GraffiTab_iOS_SDK

class HomeViewController: BackButtonViewController, CarbonTabSwipeNavigationDelegate {

    @IBOutlet weak var createBtn: UIButton!
    
    var carbonTabSwipeNavigation: CarbonTabSwipeNavigation?
    var tabs: [AnyObject]?
    var titles: [AnyObject]?
    var controllers: [UIViewController]?
    var badge: BBBadgeBarButtonItem?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        
        setupCarbonKit()
        setupButtons()
        
        configureTabBasedViews(0)
    }

    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        UIApplication.sharedApplication().setStatusBarStyle(.LightContent, animated: true)
        
        if self.navigationController!.navigationBarHidden {
            self.navigationController?.setNavigationBarHidden(false, animated: true)
        }
        
        loadUnseenNotificationsCount()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    @IBAction func onClickMenu(sender: AnyObject) {
        let mainVC = UIApplication.sharedApplication().delegate?.window??.rootViewController as! MenuContainerViewController
        mainVC.presentLeftMenuViewController()
    }
    
    @IBAction func onClickCreate(sender: AnyObject?) {
        print("CREATE")
    }
    
    @IBAction func onClickProfile(sender: AnyObject?) {
        performSegueWithIdentifier("SEGUE_PROFILE", sender: sender)
    }
    
    @IBAction func onClickNotifications(sender: AnyObject?) {
        performSegueWithIdentifier("SEGUE_NOTIFICATIONS", sender: sender)
    }
    
    @IBAction func onClickExplore(sender: AnyObject?) {
        performSegueWithIdentifier("SEGUE_EXPLORE", sender: sender)
    }
    
    @IBAction func onClickSettings(sender: AnyObject?) {
        performSegueWithIdentifier("SEGUE_SETTINGS", sender: sender)
    }
    
    // MARK: - Navigation
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "SEGUE_PROFILE" {
            let vc = segue.destinationViewController as! UserProfileViewController
            vc.user = GTSettings.sharedInstance.user
        }
    }
    
    // MARK: - Loading
    
    func loadUnseenNotificationsCount() {
        GTMeManager.getUnseenNotificationsCount({ (response) in
            self.badge?.badgeValue = response.object.stringValue
        }) { (response) in}
    }
    
    // MARK: - CarbonKitTabSwipeDelegate
    
    func carbonTabSwipeNavigation(carbonTabSwipeNavigation: CarbonTabSwipeNavigation, viewControllerAtIndex index: UInt) -> UIViewController {
        return controllers![Int(index)]
    }
    
    func carbonTabSwipeNavigation(carbonTabSwipeNavigation: CarbonTabSwipeNavigation, didMoveAtIndex index: UInt) {
        configureTabBasedViews(Int(carbonTabSwipeNavigation.currentTabIndex))
    }
    
    func configureTabBasedViews(index: Int) {   
        self.title = titles![index] as? String
    }
    
    // MARK: - Setup
    
    override func setupTopBar() {
        super.setupTopBar()
        
        let menuBtn = TintButton.init(frame: CGRectMake(0, 0, 25, 25))
        menuBtn.setImage(UIImage(named: "ic_menu_white"), forState: .Normal)
        menuBtn.tintColor = UIColor.whiteColor()
        menuBtn.addTarget(self, action: #selector(HomeViewController.onClickMenu(_:)), forControlEvents: .TouchUpInside)
        
        badge = BBBadgeBarButtonItem.init(customUIButton: menuBtn)
        badge!.shouldHideBadgeAtZero = true
        badge!.badgeOriginX = 15
        badge!.badgeOriginY = -6
        badge!.badgeBGColor = UIColor(hexString: Colors.Orange)
        badge!.badgeTextColor = UIColor.whiteColor()
        self.navigationItem.leftBarButtonItem = badge!
    }
    
    func setupCarbonKit() {
        controllers = [UIViewController]()
        controllers?.append(self.storyboard!.instantiateViewControllerWithIdentifier("FeedViewController"))
        controllers?.append(self.storyboard!.instantiateViewControllerWithIdentifier("TrendingViewController"))
        controllers?.append(self.storyboard!.instantiateViewControllerWithIdentifier("NearMeViewController"))
        controllers?.append(self.storyboard!.instantiateViewControllerWithIdentifier("RecentViewController"))
        
        tabs = [UIImage(named: "home")!, UIImage(named: "ic_whatshot_white")!, UIImage(named: "ic_near_me_white")!, UIImage(named: "ic_access_time_white")!]
        titles = ["Home", "Trending", "Near me", "Recent"]
        carbonTabSwipeNavigation = CarbonTabSwipeNavigation(items: tabs, delegate: self)
        carbonTabSwipeNavigation!.insertIntoRootViewController(self)
        
        // Styling.
        let tintColor = UIColor(hexString: Colors.Main)
        carbonTabSwipeNavigation!.setIndicatorColor(tintColor)
        carbonTabSwipeNavigation!.setTabExtraWidth(30)
        
        for (index, _) in (tabs?.enumerate())! {
            carbonTabSwipeNavigation!.carbonSegmentedControl!.setWidth(self.view.frame.width / CGFloat((tabs?.count)!), forSegmentAtIndex: index)
        }
        
        carbonTabSwipeNavigation!.setNormalColor(UIColor.blackColor().colorWithAlphaComponent(0.2))
        carbonTabSwipeNavigation!.setSelectedColor(tintColor!, font: UIFont.boldSystemFontOfSize(14))
    }
    
    func setupButtons() {
        createBtn.layer.cornerRadius = createBtn.frame.size.width / 2
        createBtn.layer.shadowRadius = 3.0
        createBtn.layer.shadowColor = UIColor.blackColor().CGColor;
        createBtn.layer.shadowOffset = CGSizeMake(1.6, 1.6)
        createBtn.layer.shadowOpacity = 0.5
        createBtn.layer.masksToBounds = false
        createBtn.backgroundColor = UIColor(hexString: Colors.Main)
        
        self.view.bringSubviewToFront(createBtn)
    }
}
